### PR TITLE
Added the third tutorial.

### PR DESCRIPTION
**Pull request by initqp**

### CODE REVIEW

**Comment by initqp** [comment]
@brucefan1983 Dear Dr. Fan, do you have any advice about this example? You can view it through [this link](https://github.com/initqp/somd_tutorials/tree/nep_part_2/3.active_learning_2). Thanks in advance!

**Comment by brucefan1983** [comment]
Looks superb. I will read this weekend.

**Comment by brucefan1983** [comment]
What are the energy scales as output from SIESTA? Like -5 eV/atom (similar to VASP) or -1000 eV/atom (similar to CP2K and QE)?  In the energy parity plot, it seems you have shifted the mean energy to zero. Is this done after the training or before it?

**Comment by brucefan1983** [comment]
So you have used `500*4*(3/4)=1500` structures for the initial training set and `500*4*(1/4)=500` for the test set. 
* At this stage, I think you don't need a test set.
* It seems this training set is quite large. In my experience, NEP is very data efficient, and you may only need 1/10 of what you used as the starting point. So `50*4=200` should be a good number for the initial training data set.
* During active learning, you can also reduce `n_max`, `neuron` and `generation`, to say `4 4`, `10`, and `50000`. 
* I would suggest use larger batchsize, preferrably fullbtach, which usually leads to the best accuracy for NEP. This is a unique property of NEP, perhaps due to the use evolutionary algorithm. It does not perform very well with mini-batches.

**Comment by brucefan1983** [comment]
With a small initial training set, you may also want to reduce the number of candidates during active learning. I think finally 1000 structures should sufficient. 

I might be wrong about the number of structures, but based on the fact that you got descent results with a batchsize of 100, I guess you don't need more than 1000 structures. If you use fullbatch training for the final data set, I believe you will get better accuracy.

**Comment by brucefan1983** [comment]
To evaluate the performance of the final model, you can predict for all the structures you generated. That is a better test than using 25% of some data set.

**Comment by initqp** [comment]
Thanks for the detailed review! I will try to answer your questions in order.

> What are the energy scales as output from SIESTA? Like -5 eV/atom (similar to VASP) or -1000 eV/atom (similar to CP2K and QE)? In the energy parity plot, it seems you have shifted the mean energy to zero. Is this done after the training or before it?

The energy scale of SIESTA is about 160 eV/atom, in between the ones of VASP and CP2K. The shifting of the energies is done after the training.

> So you have used `500*4*(3/4)=1500` structures for the initial training set and `500*4*(1/4)=500` for the test set.
> 
> * At this stage, I think you don't need a test set.
> * It seems this training set is quite large. In my experience, NEP is very data efficient, and you may only need 1/10 of what you used as the starting point. So `50*4=200` should be a good number for the initial training data set.
> * During active learning, you can also reduce `n_max`, `neuron` and `generation`, to say `4 4`, `10`, and `50000`.
> * I would suggest use larger batchsize, preferrably fullbtach, which usually leads to the best accuracy for NEP. This is a unique property of NEP, perhaps due to the use evolutionary algorithm. It does not perform very well with mini-batches.

Thanks for the suggestion. I will remove the test set from the training process. And I have modified the training parameters as you suggested. This is the new `nep_options` key:
```toml
nep_options = """
n_max      4 4
cutoff     8 6
neuron     10
lambda_v   0.0
batch      1000
generation 50000
"""
```
But I have a question about the final training: should I use larger `n_max` and `neuron` in the final training?

> With a small initial training set, you may also want to reduce the number of candidates during active learning. I think finally 1000 structures should sufficient.
> 
> I might be wrong about the number of structures, but based on the fact that you got descent results with a batchsize of 100, I guess you don't need more than 1000 structures. If you use fullbatch training for the final data set, I believe you will get better accuracy.

I will reduce the number of candidate structures to 100, thus the finial training set will only contain 1000 structures.

> To evaluate the performance of the final model, you can predict for all the structures you generated. That is a better test than using 25% of some data set.

I will modify the `copy_set.sh` script and use all visited structures as the test set in the final training.

**Comment by brucefan1983** [comment]
For the final model training, you can use larger `n_max` and `neuron`. Your original values are good.

**Comment by brucefan1983** [comment]
It is better to shift the energy (does not need to shift to 0 strictly) before training, using double precision arithmetics. This is because NEP training uses single precision throughout, and when the energy scale is larger than 100 eV/atom, there can be round-off errors. If you zoom in the energy parity plot, you may notice the data have some patterns related to a finite number of significant digits. The `nep` executable actually will give a warning message in this case. You can check the log file.

**Comment by brucefan1983** [comment]
For example, 160 000.14 meV/atom and 160 000.16 meV/atom might be represented as 160 000.1 meV/atom and 160 000.2 meV/atom, and there is thus a round-off error of the order of 0.1 meV/atom. Shifting the energy before training (with double precision operations) can solve this problem.

BTW, in your energy parity plot, the units should be eV/atom?

**Comment by initqp** [comment]
> It is better to shift the energy (does not need to shift to 0 strictly) before training, using double precision arithmetics. This is because NEP training uses single precision throughout, and when the energy scale is larger than 100 eV/atom, there can be round-off errors. If you zoom in the energy parity plot, you may notice the data have some patterns related to a finite number of significant digits. The `nep` executable actually will give a warning message in this case. You can check the log file.

OK, I get your point, this makes perfect sense to me, and I will do this within SOMD.

> For example, 160 000.14 meV/atom and 160 000.16 meV/atom might be represented as 160 000.1 meV/atom and 160 000.2 meV/atom, and there is thus a round-off error of the order of 0.1 meV/atom. Shifting the energy before training (with double precision operations) can solve this problem.
> 
> BTW, in your energy parity plot, the units should be eV/atom?

Yes it is, and this is my fault, I should correct the unit in my plots.

**Comment by initqp** [comment]
@brucefan1983 Dear Dr. Fan, I have updated this tutorial as you advised. You can view the updated version [here](https://github.com/initqp/somd_tutorials/tree/nep_part_2/3.active_learning_2). Thanks to your suggestions, the training accuracy is largely improved, and the final free energy barrier is improved by about 3 kJ/mol.

However, I still have a question about the `nep` code. During the final training, I found that the GPU memory requirement is very high (about 13 GB), thus I have to train the potential with the V100-32GB card. However, I have also tried to use two RTX3080 cards, which give out about 20 GB GPU memory in total, but `nep` still failed to launch because of the out-of-memory error.  So my question is, does `nep` constructs the full data set in every GPU card during the training? And could this be optimized in the future `nep` releases? Thanks a lot!

**Comment by brucefan1983** [comment]
> @brucefan1983 Dear Dr. Fan, I have updated this tutorial as you advised. You can view the updated version [here](https://github.com/initqp/somd_tutorials/tree/nep_part_2/3.active_learning_2). Thanks to your suggestions, the training accuracy is largely improved, and the final free energy barrier is improved by about 3 kJ/mol.
> 
> However, I still have a question about the `nep` code. During the final training, I found that the GPU memory requirement is very high (about 13 GB), thus I have to train the potential with the V100-32GB card. However, I have also tried to use two RTX3080 cards, which give out about 20 GB GPU memory in total, but `nep` still failed to launch because of the out-of-memory error. So my question is, does `nep` constructs the full data set in every GPU card during the training? And could this be optimized in the future `nep` releases? Thanks a lot!

I am glad to see that the results are improved. I would like to summarize a few points:

* During the active-learning process, using small hyperparamters (small values of `n_max`, `neuron`, `batch` which require a small value of `generation` to converge the training) can speed up the training a lot. It does not need to achieve the highest trianing accuracy at this stage. 
* In the final training, it is then suggested to increase the hyperparameters (usually a little larger than the default ones) to achieve improved accuracy. 
* You are right that `nep` constructs the full data set in every GPU card during training. Here the "full data set" should be understood as the data set for one batch. As I have suggested you to use full-batch training, this data set is large and requires a large amount of device memory. In this case, you can use a smaller batch size such that you have two batches. The code will try to divide the whole data set into two euqal-sized mini-batches. This will solve your OOM problem. From my experiences, NEP achieves the best accuracy with the full-batch mode, but it does not mean that using two or a few batches can significantly reduce the accuracy. It is only bad to use very small batch sizes. The accuracy increases with increasing batchsize in an exponential way.
* I would like to emphasize again that NEP does not need a lot of data to train. When you get the OOM problem with full-batch training, it is a sign that the training data set is still too large and might has some redundancy.
* The reason for constructing the whole data set (for a batch) in every GPU is that the multi-GPU parallelization is over the "population size", not over the structures. For example, when `population` is 50 and you use two GPUs to train, each GPU will have 25 "individuals", but each individual still requires seeing all the data.

**Comment by brucefan1983** [comment]
Another comment: do you think nuclear quantum effects can also affect your results? I have recently implemented PIMD into GPUMD, but it might be not compitable to PLUMED.

**Comment by initqp** [comment]
> > @brucefan1983 Dear Dr. Fan, I have updated this tutorial as you advised. You can view the updated version [here](https://github.com/initqp/somd_tutorials/tree/nep_part_2/3.active_learning_2). Thanks to your suggestions, the training accuracy is largely improved, and the final free energy barrier is improved by about 3 kJ/mol.
> > However, I still have a question about the `nep` code. During the final training, I found that the GPU memory requirement is very high (about 13 GB), thus I have to train the potential with the V100-32GB card. However, I have also tried to use two RTX3080 cards, which give out about 20 GB GPU memory in total, but `nep` still failed to launch because of the out-of-memory error. So my question is, does `nep` constructs the full data set in every GPU card during the training? And could this be optimized in the future `nep` releases? Thanks a lot!
> 
> I am glad to see that the results are improved. I would like to summarize a few points:
> 
> * During the active-learning process, using small hyperparamters (small values of `n_max`, `neuron`, `batch` which require a small value of `generation` to converge the training) can speed up the training a lot. It does not need to achieve the highest trianing accuracy at this stage.
> * In the final training, it is then suggested to increase the hyperparameters (usually a little larger than the default ones) to achieve improved accuracy.
> * You are right that `nep` constructs the full data set in every GPU card during training. Here the "full data set" should be understood as the data set for one batch. As I have suggested you to use full-batch training, this data set is large and requires a large amount of device memory. In this case, you can use a smaller batch size such that you have two batches. The code will try to divide the whole data set into two euqal-sized mini-batches. This will solve your OOM problem. From my experiences, NEP achieves the best accuracy with the full-batch mode, but it does not mean that using two or a few batches can significantly reduce the accuracy. It is only bad to use very small batch sizes. The accuracy increases with increasing batchsize in an exponential way.
> * I would like to emphasize again that NEP does not need a lot of data to train. When you get the OOM problem with full-batch training, it is a sign that the training data set is still too large and might has some redundancy.
> * The reason for constructing the whole data set (for a batch) in every GPU is that the multi-GPU parallelization is over the "population size", not over the structures. For example, when `population` is 50 and you use two GPUs to train, each GPU will have 25 "individuals", but each individual still requires seeing all the data.

Thanks for the detailed explanation! I understand the NEP methodology much better now. Regarding the superior expressing power of NEP, I think I should implement some (structure-based) clustering algorithms in active learning to further reduce the training set size. 



> Another comment: do you think nuclear quantum effects can also affect your results? I have recently implemented PIMD into GPUMD, but it might be not compitable to PLUMED.


I am sorry about this, but I know little about the PIMD method. The only thing I know is that the bias potential introduced by PLUMED should act on the centroid "system", e.g., as shown in [this article](https://pubs.acs.org/action/showCitFormats?doi=10.1021/acs.jctc.2c00874&ref=pdf). Regarding the current example, I think the NQE is not very important, since the reaction coordinate does not contain the positions of any hydrogen atom.

**Comment by brucefan1983** [comment]
Thanks for the reference, which is very useful. 

My last comment is that you don't need try to use 2 GPUs to run GPUMD for small boxes. In your case, the box length is even smaller than twice of the cutoff radius and GPUMD will use the so-called "small-box" version of NEP to run it. In this case it will always use one GPU only. If your box length is between 2.5 and 10 times of the cutoff radius, GPUMD will refuse to run with multi-GPUs because there is not much gain of speed.

Multi-GPU is only benificial for very large systems (such as > 1e5 atoms).

**Comment by initqp** [comment]
> Thanks for the reference, which is very useful.
> 
> My last comment is that you don't need try to use 2 GPUs to run GPUMD for small boxes. In your case, the box length is even smaller than twice of the cutoff radius and GPUMD will use the so-called "small-box" version of NEP to run it. In this case it will always use one GPU only. If your box length is between 2.5 and 10 times of the cutoff radius, GPUMD will refuse to run with multi-GPUs because there is not much gain of speed.
> 
> Multi-GPU is only benificial for very large systems (such as > 1e5 atoms).

Thanks a lot, this is very practical advice, maybe I should add this to the main text.

**Comment by initqp** [comment]
@brucefan1983  I think it is time to merge this, if you have any other advice or question, please feel free to comment under this thread and/or the Discussions page :)

**Comment by brucefan1983** [comment]
I do not meant it is not good to use multi GPUs to train. I only meant this for running MD. 

For NEP training, using `N` GPUs will almost always make it `N` times faster! You should have observed this if you have successfully used multi GPUs for training.

**Comment by initqp** [comment]
> I do not meant it is not good to use multi GPUs to train. I only meant this for running MD.
> 
> For NEP training, using `N` GPUs will almost always make it `N` times faster! You should have observed this if you have successfully used multi GPUs for training.

Really sorry for the inconvenience! It's a very stupid misunderstanding. I will revert these commits.
